src/test/clone/test_clone.rs

```diff
@@ -0,0 +1,187 @@
+use core::ffi::c_void;
+use std::error::Error;
+use std::sync::atomic::{self, AtomicU32};
+
+use linux_api::ldt::linux_user_desc;
+use linux_api::sched::CloneFlags;
+use rustix::mm::{MapFlags, MprotectFlags, ProtFlags};
+use rustix::thread::{futex, FutexFlags, FutexOperation};
+use rustix::time::Timespec;
+use test_utils::TestEnvironment as TestEnv;
+use test_utils::{running_in_shadow, set, ShadowTest};
+use vasi_sync::lazy_lock::LazyLock;
+use vasi_sync::scchannel::SelfContainedChannel;
+
+const CLONE_TEST_STACK_NBYTES: usize = 4 * 4096;
+
+// TODO: consider adding this type to `linux-api` and importing from there for
+// consistency.  I ran into trouble doing so due to *cbindgen* failing to parse
+// the bitfield definition created by *bindgen*. This problem goes away if/when
+// we stop re-exporting `linux-api` to C.
+fn make_empty_tls() -> linux_user_desc {
+    let mut desc: linux_user_desc = unsafe { core::mem::zeroed() };
+    desc.set_seg_not_present(1);
+    desc.set_read_exec_only(1);
+    desc
+}
+
+struct ThreadStack {
+    base: *mut c_void,
+    size: usize,
+}
+
+impl ThreadStack {
+    pub fn new(size: usize) -> Self {
+        let base = unsafe {
+            rustix::mm::mmap_anonymous(
+                core::ptr::null_mut(),
+                CLONE_TEST_STACK_NBYTES,
+                ProtFlags::READ | ProtFlags::WRITE,
+                MapFlags::PRIVATE | MapFlags::STACK,
+            )
+        }
+        .unwrap();
+
+        // Use first page as a guard page.
+        unsafe { rustix::mm::mprotect(base, 4096, MprotectFlags::empty()) }.unwrap();
+
+        Self { base, size }
+    }
+
+    pub fn top(&self) -> *mut c_void {
+        unsafe { self.base.add(self.size) }
+    }
+}
+
+impl Drop for ThreadStack {
+    fn drop(&mut self) {
+        unsafe { rustix::mm::munmap(self.base, self.size) }.unwrap()
+    }
+}
+
+fn test_clone_minimal() -> Result<(), Box<dyn Error>> {
+    static THREAD_DONE_CHANNEL: LazyLock<SelfContainedChannel<()>> =
+        LazyLock::const_new(SelfContainedChannel::new);
+    extern "C" fn thread_fn(_param: *mut c_void) -> i32 {
+        THREAD_DONE_CHANNEL.send(());
+        0
+    }
+    let mut tls = make_empty_tls();
+    let stack = ThreadStack::new(CLONE_TEST_STACK_NBYTES);
+    let flags = CloneFlags::CLONE_VM
+        | CloneFlags::CLONE_FS
+        | CloneFlags::CLONE_FILES
+        | CloneFlags::CLONE_SIGHAND
+        | CloneFlags::CLONE_THREAD
+        | CloneFlags::CLONE_SYSVSEM
+        | CloneFlags::CLONE_SETTLS;
+    let child = unsafe {
+        libc::clone(
+            thread_fn,
+            stack.top(),
+            flags.bits().try_into().unwrap(),
+            core::ptr::null_mut(),
+            core::ptr::null_mut::<i32>(),
+            &mut tls,
+        )
+    };
+    assert!(child > 0);
+    THREAD_DONE_CHANNEL.receive().unwrap();
+    Ok(())
+}
+
+fn test_clone_clear_tid() -> Result<(), Box<dyn Error>> {
+    extern "C" fn thread_fn(_param: *mut c_void) -> i32 {
```

Comment:
Should the thread function for each test be documented to say that can't call libc functions, as in the commit comment?

Reply:
Done